File: python/jobctl.py
"""jobctl — the Python half of the cross-language conformance test.

Same commands as the Go jobctl, against the same directory. Neither knows the
other exists; both only know the record.
"""

import argparse
import json
import os
import platform
import sys
from datetime import datetime, timezone

try:
    import abstraction_config
except ImportError:
    sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), os.pardir, os.pardir, "abstraction-config", "python"))
    import abstraction_config

from abstraction_job import (
    CANCEL,
    PAUSE,
    RUN,
    FileStore,
    JobError,
    Record,
    parse_opaque,
)


def store_root():
    """Resolve a store root rather than demanding one.

    JOB_STORE is jobctl's own override and stays here: the conformance harness
    points three implementations at one directory with it and must not inherit
    what the machine has configured. Every other rung is the config layer's
    answer, in one place, for all three languages.
    """
    value = os.environ.get("JOB_STORE")
    if value:
        return value, "JOB_STORE"
    try:
        return abstraction_config.job_store()
    except OSError as e:
        sys.exit("jobctl: %s\n  ABSTRACTION_STORE=…   names one" % e)


def store() -> FileStore:
    root, came_from = store_root()
    try:
        return FileStore(root)
    except OSError as e:
        sys.exit(
            "jobctl: %s: %s\n"
            "  that root came from %s\n"
            "  jobd setup --show     what this machine has configured, and which file said so\n"
            "  ABSTRACTION_STORE=…   names a store for one run" % (root, e, came_from)
        )


def main() -> None:
    # Windows text mode would turn every LF into CRLF, and a record printed here
    # is a conformance surface compared against Go and C++ byte for byte. Without
    # this the three implementations agree about every field and the comparison
    # still fails -- on line endings, which is the least interesting way to
    # disagree and the hardest to see in a diff.
    #
    # The encoding is here for the same reason. A record spells a filename in
    # raw UTF-8 [JOB-E6], and a Windows console defaults to the machine's code
    # page, so printing an accented name raised UnicodeEncodeError from a tool
    # that had read the record perfectly well.
    try:
        sys.stdout.reconfigure(newline="\n", encoding="utf-8")
    except AttributeError:  # pragma: no cover - Python < 3.7
        pass

    p = argparse.ArgumentParser(prog="jobctl.py")
    sub = p.add_subparsers(dest="cmd", required=True)

    s_submit = sub.add_parser("submit")
    s_submit.add_argument("--kind", required=True, help="what this job is; who can read the spec")
    s_submit.add_argument("--spec", required=True, help="the job's spec, as raw JSON")
    s_submit.add_argument("--total", type=int, default=0, help="total work, in the kind's own units")

    s_claim = sub.add_parser("claim")
    s_claim.add_argument("id")
    s_claim.add_argument("--owner", required=True)
    s_claim.add_argument("--ttl", type=float, default=30.0)

    s_prog = sub.add_parser("progress")
    s_prog.add_argument("id")
    s_prog.add_argument("--epoch", type=int, required=True)
    s_prog.add_argument("--done", type=int, default=0)
    s_prog.add_argument("--checkpoint", default="", help="what a successor needs to resume, as raw JSON")

    s_fin = sub.add_parser("finish")
    s_fin.add_argument("id")
    s_fin.add_argument("--epoch", type=int, required=True)
    s_fin.add_argument("--state", default="transferred")

    s_show = sub.add_parser("show")
    s_show.add_argument("id")

    # Say what should happen, without holding the lease. No --epoch, and that
    # absence is the feature: the caller is not the worker.
    s_int = sub.add_parser("intent")
    s_int.add_argument("id")
    s_int.add_argument("want", choices=[RUN, PAUSE, CANCEL])
    s_int.add_argument("--by", default="")

    # Ask the holder for the lease back. --epoch is the one the caller SAW, not
    # one it holds: a third party recalling a residency it has only read.
    s_rec = sub.add_parser("recall")
    s_rec.add_argument("id")
    s_rec.add_argument("--epoch", type=int, required=True)
    s_rec.add_argument("--reason", required=True)
    s_rec.add_argument("--grace", type=float, default=30.0)
    s_rec.add_argument("--by", default="")

    sub.add_parser("orphans")
    sub.add_parser("list")

    a = p.parse_args()
    st = store()

    try:
        if a.cmd == "submit":
            # Deliberately ignorant of what a job IS: the spec goes through
            # untouched, which is the same contract the module itself keeps.
            rec = Record(id="", kind=a.kind, spec=parse_opaque(a.spec))
            rec.progress.total = a.total
            print(st.submit(rec))

        elif a.cmd == "claim":
            r = st.claim(a.id, a.owner, a.ttl)
            cp = json.dumps(r.checkpoint, separators=(",", ":")) if r.checkpoint else "none"
            print(f"epoch={r.lease.epoch} state={r.state} checkpoint={cp}")

        elif a.cmd == "progress":
            def mutate(r):
                r.progress.done = a.done
                r.progress.updated_at = datetime.now(timezone.utc)
                if a.checkpoint:
                    r.checkpoint = parse_opaque(a.checkpoint)

            r = st.update(a.id, a.epoch, mutate)
            cp = json.dumps(r.checkpoint, separators=(",", ":")) if r.checkpoint else "none"
            print(f"done={r.progress.done} checkpoint={cp}")

        elif a.cmd == "finish":
            def mutate(r):
                r.state = a.state

            r = st.update(a.id, a.epoch, mutate)
            print(f"state={r.state}")

        elif a.cmd == "show":
            sys.stdout.write(st.load(a.id).to_json().decode())

        elif a.cmd == "intent":
            by = a.by or f"jobctl.py@{platform.node()}:{os.getpid()}"
            r = st.set_intent(a.id, a.want, by)
            print(f"{r.id} {r.wants()}")

        elif a.cmd == "recall":
            by = a.by or f"jobctl.py@{platform.node()}:{os.getpid()}"
            r = st.recall(a.id, a.epoch, a.reason, by, a.grace)
            print(f"{r.id} recalled until {r.lease.recall.until.isoformat()}")

        elif a.cmd == "orphans":
            for r in st.orphans():
                cp = json.dumps(r.checkpoint, separators=(",", ":")) if r.checkpoint else "none"
                print(f"{r.id} kind={r.kind} state={r.state} checkpoint={cp}")

        elif a.cmd == "list":
            for r in st.list():
                cp = json.dumps(r.checkpoint, separators=(",", ":")) if r.checkpoint else "none"
                print(
                    f"{r.id} kind={r.kind} state={r.state} "
                    f"done={r.progress.done} checkpoint={cp}"
                )

    except JobError as e:
        sys.exit(f"jobctl: {type(e).__name__}: {e}")


if __name__ == "__main__":
    main()

File: cpp/cmd/jobctl.cpp
// jobctl drives the job store from a shell, so the cross-language conformance
// test can be an actual script that runs three implementations against one
// directory rather than a mock of one talking to a mock of another.
//
// Deliberately ignorant of what a job IS: it takes a kind and a spec as raw
// JSON and never looks inside, which is the same contract the library keeps.

#include <cstdlib>
#include <iostream>
#include <abstraction/job/store.h>
#include <map>
#include <string>
#include <vector>

#ifdef _WIN32
#include <fcntl.h>
#include <io.h>
#endif

namespace {

using abstraction::job::FileStore;
using abstraction::job::Json;
using abstraction::job::Record;

struct Args {
    std::string id;  // the first non-flag argument, in any position
    // Every non-flag argument after the id. `intent <id> <want>` needs a
    // second one, and dropping it silently would have made the want default
    // to whatever the handler chose rather than what the caller typed.
    std::vector<std::string> positional;
    std::map<std::string, std::string> flags;
};

Args parse_args(const std::vector<std::string>& argv) {
    Args a;
    for (std::size_t i = 0; i < argv.size(); ++i) {
        const std::string& token = argv[i];
        if (token.size() > 1 && token[0] == '-') {
            std::string name = token.substr(token[1] == '-' ? 2 : 1);
            const auto eq = name.find('=');
            if (eq != std::string::npos) {
                a.flags[name.substr(0, eq)] = name.substr(eq + 1);
                continue;
            }
            if (i + 1 < argv.size()) {
                a.flags[name] = argv[++i];
            } else {
                a.flags[name] = "";
            }
            continue;
        }
        if (a.id.empty()) {
            a.id = token;
        } else {
            a.positional.push_back(token);
        }
    }
    return a;
}

std::string flag(const Args& a, const std::string& name, const std::string& fallback = "") {
    const auto it = a.flags.find(name);
    return it == a.flags.end() ? fallback : it->second;
}

// Puts the checkpoint on one line. The record on disk is indented for humans,
// so a checkpoint read back out carries newlines — and this output is a
// conformance surface a harness parses, where "same value, different
// whitespace" counts as two implementations disagreeing.
std::string compact(const Record& r) {
    return r.checkpoint ? r.checkpoint->dump() : "none";
}

[[noreturn]] void fatal(const std::string& message) {
    std::cerr << "jobctl: " << message << std::endl;
    std::exit(1);
}

[[noreturn]] void usage() {
    std::cout << "usage: jobctl <submit|claim|progress|finish|show|intent|recall|orphans> [args]"
                 "   (JOB_STORE must be set)\n";
    std::cout << "  submit --kind K --spec '<json>' [--total N] [--requires a,b]\n";
    std::exit(2);
}

std::chrono::milliseconds ttl_from(const std::string& raw, double fallback_seconds) {
    if (raw.empty()) {
        return std::chrono::milliseconds(static_cast<long long>(fallback_seconds * 1000.0));
    }
    // Seconds as a plain number, matching Go and Python. A Go-style duration
    // suffix is accepted too, because an older harness may still pass "2s".
    std::size_t consumed = 0;
    double value = 0.0;
    try {
        value = std::stod(raw, &consumed);
    } catch (const std::exception&) {
        fatal("invalid --ttl \"" + raw + "\"");
    }
    const std::string suffix = raw.substr(consumed);
    double multiplier = 1000.0;
    if (suffix == "ms") {
        multiplier = 1.0;
    } else if (suffix == "m") {
        multiplier = 60000.0;
    } else if (suffix == "h") {
        multiplier = 3600000.0;
    } else if (!suffix.empty() && suffix != "s") {
        fatal("invalid --ttl \"" + raw + "\"");
    }
    return std::chrono::milliseconds(static_cast<long long>(value * multiplier));
}

Json parse_json_flag(const std::string& raw, const char* what) {
    try {
        return Json::parse(raw);
    } catch (const Json::Error& e) {
        fatal(std::string(what) + " is not valid JSON: " + e.what());
    }
}

std::vector<std::string> split_commas(const std::string& raw) {
    std::vector<std::string> out;
    std::size_t start = 0;
    for (;;) {
        const auto comma = raw.find(',', start);
        if (comma == std::string::npos) {
            out.push_back(raw.substr(start));
            return out;
        }
        out.push_back(raw.substr(start, comma - start));
        start = comma + 1;
    }
}

}  // namespace

int main(int argc, char** argv) {
#ifdef _WIN32
    // `show` must emit the record's own bytes. Text mode would turn every LF in
    // the canonical encoding into CRLF, so a harness diffing this output against
    // the file on disk would see two implementations disagreeing over nothing.
    _setmode(_fileno(stdout), _O_BINARY);
#endif
    if (argc < 2) {
        usage();
    }
    const std::string command = argv[1];
    const std::vector<std::string> rest(argv + 2, argv + argc);
    const Args args = parse_args(rest);

    const char* root = std::getenv("JOB_STORE");
    if (root == nullptr || *root == '\0') {
        fatal("JOB_STORE is not set");
    }

    try {
        FileStore store(root);

        if (command == "submit") {
            Record r;
            r.kind = flag(args, "kind");
            r.spec = parse_json_flag(flag(args, "spec", "{}"), "--spec");
            const std::string total = flag(args, "total");
            if (!total.empty()) {
                r.progress.total = std::stoll(total);
            }
            const std::string requires_flag = flag(args, "requires");
            if (!requires_flag.empty()) {
                r.requires_capabilities = split_commas(requires_flag);
            }
            std::cout << store.submit(std::move(r)) << std::endl;

        } else if (command == "claim") {
            if (args.id.empty()) {
                fatal("claim needs a job id");
            }
            const Record r =
                store.claim(args.id, flag(args, "owner"), ttl_from(flag(args, "ttl"), 30.0));
            // The epoch and the predecessor's checkpoint are what a new owner needs.
            std::cout << "epoch=" << r.lease.epoch << " state=" << r.state
                      << " checkpoint=" << compact(r) << std::endl;

        } else if (command == "progress") {
            if (args.id.empty()) {
                fatal("progress needs a job id");
            }
            const std::string done = flag(args, "done", "0");
            const std::string checkpoint = flag(args, "checkpoint");
            const Record r = store.update(
                args.id, std::stoll(flag(args, "epoch", "0")), [&](Record& rec) {
                    rec.progress.done = std::stoll(done);
                    rec.progress.updated_at = abstraction::job::Clock::now();
                    if (!checkpoint.empty()) {
                        rec.checkpoint = parse_json_flag(checkpoint, "--checkpoint");
                    }
                });
            std::cout << "done=" << r.progress.done << " checkpoint=" << compact(r) << std::endl;

        } else if (command == "finish") {
            if (args.id.empty()) {
                fatal("finish needs a job id");
            }
            const std::string wanted = flag(args, "state", abstraction::job::state::kTransferred);
            const Record r =
                store.update(args.id, std::stoll(flag(args, "epoch", "0")), [&](Record& rec) {
                    if (!abstraction::job::is_valid_state(wanted)) {
                        throw abstraction::job::Invalid("invalid state \"" + wanted + "\"");
                    }
                    rec.state = wanted;
                });
            std::cout << "state=" << r.state << std::endl;

        } else if (command == "show") {
            if (args.id.empty()) {
                fatal("show needs a job id");
            }
            std::cout << store.load(args.id).encode() << std::flush;

        } else if (command == "intent") {
            // A command rather than a flag on claim, because the whole point is
            // that the caller is NOT the worker: no epoch is presented, and
            // none is needed.
            if (args.id.empty() || args.positional.empty()) {
                fatal("usage: jobctl intent <id> <run|pause|cancel> [--by who]");
            }
            const std::string by = flag(args, "by", "jobctl-cpp");
            const Record r = store.set_intent(args.id, args.positional.front(), by);
            std::cout << r.id << " " << r.wants() << std::endl;

        } else if (command == "recall") {
            // --epoch is the one the caller SAW, not one it holds: a third
            // party recalling a residency it has only read.
            if (args.id.empty() || flag(args, "reason").empty()) {
                fatal("usage: jobctl recall <id> --epoch N --reason WHY [--grace SECONDS] [--by who]");
            }
            const Record r = store.recall(
                args.id, std::stoll(flag(args, "epoch", "0")), flag(args, "reason"),
                flag(args, "by", "jobctl-cpp"), ttl_from(flag(args, "grace"), 30.0));
            std::cout << r.id << " recalled until " << abstraction::job::format_rfc3339(r.lease.recall->until)
                      << std::endl;

        } else if (command == "orphans") {
            for (const Record& r : store.orphans()) {
                std::cout << r.id << " kind=" << r.kind << " state=" << r.state
                          << " checkpoint=" << compact(r) << std::endl;
            }

        } else {
            usage();
        }
    } catch (const abstraction::job::JobError& e) {
        fatal(std::string(e.name()) + ": " + e.what());
    } catch (const std::exception& e) {
        fatal(e.what());
    }
    return 0;
}
